TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Service/NodeTreeService.php

```diff
@@ -0,0 +1,52 @@
+<?php
+namespace TYPO3\TYPO3CR\Service;
+
+/*
+ * This file is part of the TYPO3.TYPO3CR package.
+ *
+ * (c) Contributors of the Neos Project - www.neos.io
+ *
+ * This package is Open Source Software. For the full copyright and license
+ * information, please view the LICENSE file which was distributed with this
+ * source code.
+ */
+
+use TYPO3\Flow\Annotations as Flow;
+use TYPO3\TYPO3CR\Domain\Factory\NodeFactory;
+use TYPO3\TYPO3CR\Domain\Model\NodeInterface;
+
+/**
+ * The node tree service
+ * @Flow\Scope("singleton")
+ */
+class NodeTreeService
+{
+
+    /**
+     * @Flow\Inject
+     * @var NodeFactory
+     */
+    protected $nodeFactory;
+
+
+    /**
+     * @param NodeInterface $node
+     * @param callable $action
+     * @return void
+     */
+    public function traverseTree(NodeInterface $node, callable $action)
+    {
+        $continue = $action($node);
+        if ($continue !== false && $node->hasChildNodes()) {
+            foreach ($node->getChildNodes() as $childNode) {
+                $this->traverseTree($childNode, $action);
+            }
+        }
+        $this->nodeFactory->removeNodeFromCache($node);
```

Comment:
Errr, you're totally right. Was a bit confused...

The traversal "algorithm" should do fine with reset, but I'm a bit reluctant to use ist. I'd prefer completely deactivating the caches during tree traversal, because:
a) With write/reset we then have n cache write operations, 0 cache hits and n reset operations for n nodes. Though everything happens in-memory and thus is fast, it's unnecessary none the less
b) Peak memory usage _should_ decrease further, especially when there are nodes with lots of child nodes in the tree. I'll have to provide some numbers here, though, I guess.
c) (Just a minor issue performance-wise) Resetting the caches might throw out nodes that are already happily residing there without causing any trouble. Think of node:repair for example, where the tree traversal is just one of many operations.

The two important questions before I get started:
Would introducing something like disableCache() and enableCache() in NodeFactory and ContentContext be okay?
Why does the NodeFactory not use the ContentContext's first level node cache in the first place? Would make stuff a little easier here if it would.

Maybe we should talk this through directly via hangout